airflow/models/dagcode.py

```diff
@@ -0,0 +1,108 @@
+from airflow.utils.sqlalchemy import UtcDateTime
+
+log = logging.getLogger(__name__)
+
+
+class DagCodeModel(Base):
+    """A table for DAGs code.
+
+    dag_code table contains code of DAG files synchronized by scheduler.
+    This feature is controlled by:
+
+    * ``[core] store_serialized_dags = True``: enable this feature
+    * ``[core] store_code = True``: enable this feature
+
+    For details on dag serialization see SerializedDagModel
+    """
+    __tablename__ = 'dag_code'
+
+    fileloc = Column(String(2000), primary_key=True)
+    # The max length of fileloc exceeds the limit of indexing.
+    fileloc_hash = Column(Integer, primary_key=True)
+    last_updated = Column(UtcDateTime, nullable=False)
+    source_code = Column(Text(), nullable=False)
+
+    __table_args__ = (
+        Index('idx_fileloc_code_hash', fileloc_hash, unique=False),
+    )
+
+    def __init__(self, full_filepath: str):
+        self.fileloc = full_filepath
+        self.fileloc_hash = DagCodeModel.dag_fileloc_hash(self.fileloc)
+        self.last_updated = timezone.utcnow()
+        self.source_code = DagCodeModel._read_code(self.fileloc)
+
+    @classmethod
+    def _read_code(cls, fileloc: str):
+        try:
+            with open(fileloc, 'r') as source:
+                source_code = source.read()
+        except IOError:
+            source_code = "Couldn't read source file {}.".format(fileloc)
+        return source_code
+
+    @provide_session
+    def write_code(self, session=None):
+        """Writes code into database.
+
+        :param session: ORM Session
+        """
+        session.merge(self)
+
+    @classmethod
+    @provide_session
+    def remove_deleted_code(cls, alive_dag_filelocs: List[str], session=None):
+        """Deletes code not included in alive_dag_filelocs.
+
+        :param alive_dag_filelocs: file paths of alive DAGs
+        :param session: ORM Session
+        """
+        alive_fileloc_hashes = [
+            cls.dag_fileloc_hash(fileloc) for fileloc in alive_dag_filelocs]
+
+        log.debug("Deleting code from %s table ", cls.__tablename__)
+
+        session.execute(
+            cls.__table__.delete().where(
+                and_(cls.fileloc_hash.notin_(alive_fileloc_hashes),
+                     cls.fileloc.notin_(alive_dag_filelocs))))
+
+    @classmethod
+    def dag_fileloc_hash(cls, full_filepath: str) -> int:
+        """"Hashing file location for indexing.
+
+        :param full_filepath: full filepath of DAG file
+        :return: hashed full_filepath
+        """
+        # hashing is needed because the length of fileloc is 2000 as an Airflow convention,
+        # which is over the limit of indexing. If we can reduce the length of fileloc, then
+        # hashing is not needed.
+        import hashlib
+        return int.from_bytes(
+            hashlib.sha1(
+                full_filepath.encode('utf-8')).digest()[-2:],
+            byteorder='big', signed=False)
```

Comment:
Since we are using this in a PK we should use more than one hex-nibble to (greatly) reduce the risk of collisions.

```
        hash = hashlib.sha1(full_filepath.encode('utf-8')).digest()
        return int.from_bytes(
            hash[-4:],
            byteorder='big', signed=False)
```

(This goes from 0...64k possible values to 0..4billion possible values.)

Buut this does possibly mean we need to migrate the fileloc_hash value on serialized_dag table. Dang.

Reply:
I have decided to add id column and create uniqness constraint on fileloc instead.
It sounds like a valid suggestion though.
Maybe in another PR?

Reply:
I agree. I shall create this migration